clients/src/main/java/org/apache/kafka/clients/NetworkClient.java

```diff
@@ -516,13 +516,15 @@ private void initiateConnect(Node node, long now) {
         /* true iff there is a metadata request that has been sent and for which we have not yet received a response */
         private boolean metadataFetchInProgress;
 
-        /* the last timestamp when no broker node is available to connect */
-        private long lastNoNodeAvailableMs;
```

Comment:
You're right.
Even though maybeUpdate() returns the maximum time to wait as a backoff, network event can cause immediate progress to the next step.
I basically agree with your four steps except one thing about 2.
So whenever `leastLoadedNode` returns null, that means all nodes are currently backing off right? In case, we should return the minimum remaining reconnect backoff time of all nodes I guess. I will add an followup commit based on this. let's make further discussion on updated patch if necessary.

Thanks for nice catch.